src/misc_early.rs

```diff
@@ -113,4 +124,43 @@ impl EarlyLintPass for MiscEarly {
             }
         }
     }
+
+    fn check_expr(&mut self, cx: &EarlyContext, expr: &Expr) {
+        if let ExprKind::Call(ref paren, _) = expr.node {
+            if let ExprKind::Paren(ref closure) = paren.node {
+                if let ExprKind::Closure(_, ref decl, ref block) = closure.node {
+                    span_lint_and_then(cx,
+                                       REDUNDANT_CLOSURE_CALL,
+                                       expr.span,
+                                       "Try not to call a closure in the expression where it is declared.",
+                                       |db| {
+                                            if decl.inputs.len() == 0 {
+                                                let hint = format!("{}", snippet(cx, block.span, ".."));
+                                                db.span_suggestion(expr.span, "Try doing something like: ", hint);
+                                            }
+                                        });
+                }
+            }
+        }
+    }
+
+    fn check_block(&mut self, cx: &EarlyContext, block: &Block) {
+        for w in block.stmts.windows(2) {
+            if_let_chain! {[
+                let StmtKind::Decl(ref first, _) = w[0].node,
+                let DeclKind::Local(ref local) = first.node,
+                let Option::Some(ref t) = local.init,
+                let ExprKind::Closure(_,_,_) = t.node,
+                let PatKind::Ident(_,sp_ident,_) = local.pat.node,
+                let StmtKind::Semi(ref second,_) = w[1].node,
+                let ExprKind::Assign(_,ref call) = second.node,
+                let ExprKind::Call(ref closure,_) = call.node,
+                let ExprKind::Path(_,ref path) = closure.node
+            ], {
+                if sp_ident.node == (&path.segments[0]).identifier {
+                    span_lint(cx, REDUNDANT_CLOSURE_CALL, second.span, "Closure called just once immediately after it was declared");
```

Comment:
There's a false positive here, it's possible that it might be called again. However this involves ExprUseVisitor, I'll fix that in a PR later.